bevy_reflect: Recursive registration

MrGVSV opened the pull request:
# Objective

Resolves #4154

Currently, registration must all be done manually:

```rust
#[derive(Reflect)]
struct Foo(Bar);

#[derive(Reflect)]
struct Bar(Baz);

#[derive(Reflect)]
struct Baz(usize);

fn main() {
  // ...
  app
    .register_type::<Foo>()
    .register_type::<Bar>()
    .register_type::<Baz>()
    // .register_type::<usize>() <- This one is handled by Bevy, thankfully
  // ...
}
```

This can grow really quickly and become very annoying to add, remove, and update as types change. It would be great if we could help reduce the number of types that a user must manually implement themselves.

## Solution

As suggested in #4154, this PR adds automatic recursive registration. Essentially, when a type is registered, it may now also choose to register additional types along with it using the new `GetTypeRegistration::register_type_dependencies` trait method.

The `Reflect` derive macro now automatically does this for all fields in structs, tuple structs, struct variants, and tuple variants. This is also done for tuples, arrays, `Vec<T>`, `HashMap<K, V>`, and `Option<T>`.

This allows us to simplify the code above like:

```rust
#[derive(Reflect)]
struct Foo(Bar);

#[derive(Reflect)]
struct Bar(Baz);

#[derive(Reflect)]
struct Baz(usize);

fn main() {
  // ...
  app.register_type::<Foo>()
  // ...
}
```

This automatic registration only occurs if the type has not yet been registered. If it has been registered, we simply skip it and move to the next one. This reduces the cost of registration and prevents overwriting customized registrations.

## Considerations

While this does improve ergonomics on one front, it's important to look at some of the arguments against adopting a PR like this.

#### Generic Bounds

~~Since we need to be able to register the fields individually, we need those fields to implement `GetTypeRegistration`. This forces users to then add this trait as a bound on their generic arguments. This annoyance could be relieved with something like #5772.~~

This is no longer a major issue as the `Reflect` derive now adds the `GetTypeRegistration` bound by default. This should technically be okay, since we already add the `Reflect` bound.

However, this can also be considered a breaking change for manual implementations that left out a `GetTypeRegistration` impl ~~or for items that contain dynamic types (e.g. `DynamicStruct`) since those also do not implement `GetTypeRegistration`~~.

#### Registration Assumptions

By automatically registering fields, users might inadvertently be relying on certain types to be automatically registered. If `Foo` auto-registers `Bar`, but `Foo` is later removed from the code, then anywhere that previously used or relied on `Bar`'s registration would now fail.

---

## Changelog

- Added recursive type registration to structs, tuple structs, struct variants, tuple variants, tuples, arrays, `Vec<T>`, `HashMap<K, V>`, and `Option<T>`
- Added a new trait in the hidden `bevy_reflect::__macro_exports` module called `RegisterForReflection`
- Added `GetTypeRegistration` impl for `bevy_render::render_asset::RenderAssetUsages`

## Migration Guide

All types that derive `Reflect` will now automatically add `GetTypeRegistration` as a bound on all (unignored) fields. This means that all reflected fields will need to also implement `GetTypeRegistration`.

If all fields **derive** `Reflect` or are implemented in `bevy_reflect`, this should not cause any issues. However, manual implementations of `Reflect` that excluded a `GetTypeRegistration` impl for their type will need to add one.

```rust
#[derive(Reflect)]
struct Foo<T: FromReflect> {
  data: MyCustomType<T>
}

// OLD
impl<T: FromReflect> Reflect for MyCustomType<T> {/* ... */}

// NEW
impl<T: FromReflect + GetTypeRegistration> Reflect for MyCustomType<T> {/* ... */}
impl<T: FromReflect + GetTypeRegistration> GetTypeRegistration for MyCustomType<T> {/* ... */}
```

MrGVSV commented on crates/bevy_reflect/src/type_registry.rs:
```diff
@@ -83,6 +94,21 @@ impl TypeRegistry {
+        T::register_type_dependencies(self);
+    }
+
+    /// Attempts to register the type `T` if it has not yet been registered already.
+    ///
+    /// If the registration for type `T` already exists, it will not be registered again.
+    ///
+    /// To register the type, overwriting any existing registration, use [register](Self::register) instead.
+    pub fn try_register<T>(&mut self)
+    where
+        T: GetTypeRegistration + 'static,
+    {
+        if !self.contains(TypeId::of::<T>()) {
+            self.register::<T>();
+        }
     }
```
I would too but that might be confusing if `app.register()` returns `&mut App` :/

MrGVSV replied:
Oh you know what, I’m mixing up my PRs. It's too early in the morning I guess 😅

Yeah you might be right about this. I'll try it out and ensure things still work correctly.

MrGVSV replied:
This was addressed by your PR (#6487), but was not accounted for in this PR. So I updated the code based on those changes.

MrGVSV commented on examples/reflection/generic_reflection.rs:
```diff
@@ -13,7 +14,7 @@ fn main() {
 
 #[derive(Reflect)]
-struct MyType<T: Reflect> {
+struct MyType<T: Reflect + GetTypeRegistration> {
```
This is why I’d really like #5772 to get merged. It would combine all these core traits into one that does not need to worry about object-safety

MrGVSV replied:
Yeah but it means we keep it to two traits that a user mist know, even as we add more and more core traits.

Separating traits is also about readability imo so if we did combine everything then we'd eventually end up with a potentially massive trait

MrGVSV commented:
We might actually be able to improve the usability here by adding `GetTypeRegistration` to the `where` clause of the impl like we do with `Reflect` 🤔

wilk10 commented:
I can't speak to the actual implementation but as a user I have found it quite painful without this recursive registration, so this is a very welcome change, thanks a lot! Is there any chance that it makes it into 0.11?

I understand the concern that when a outer type is removed, the inner types won't be automatically registered any more. But in this case there's at least a warning thrown by Bevy, so that does give some guidance and should mitigate the issue, imo.

MrGVSV commented:
I'm actually realizing this might not be in an ideal state yet thanks to https://github.com/bevyengine/bevy/issues/8965. This will prevent users from including dynamic types in the fields of their reflected items (since they do not implement `GetTypeRegistration`). For example, this will fail to compile:

```rust
#[derive(Reflect)]
struct Foo {
  data: DynamicStruct
}
```

On the other hand, this is an issue anyways since this problem slipped through the cracks in #6056.

If this isn't that big a deal, we can move forward with it. Otherwise, I think we really need to rethink how we add type bounds.

SkiFire13 commented on crates/bevy_reflect/src/type_registry.rs:
```diff
@@ -92,24 +99,53 @@ impl TypeRegistry {
         T: GetTypeRegistration,
     {
-        self.add_registration(T::get_type_registration());
+        if !self.add_registration(T::get_type_registration()) {
```
This could avoid computing `T::get_type_registration` is `T` is already registered (assuming `T: GetTypeRegistration always returns a `TypeRegistration` for `T` and not for other types)

alice-i-cecile commented:
@MrGVSV, could you resolve conflicts? This is the only remaining reflection PR in the milestone and you and @soqb have said you'd like to land this in 0.12.

MrGVSV commented:
I updated the Migration Guide:

> Deriving `Reflect` now requires that all active fields also implement `GetTypeRegistration` (which is automatically implemented when deriving `Reflect`). This means that types which do not implement `GetTypeRegistration` now need to be ignored:
> 
> ```rust
> // OLD
> #[derive(Reflect)]
> #[reflect(from_reflect = false)]
> struct Foo {
>   data: DynamicStruct
> }
> 
> // NEW
> #[derive(Reflect)]
> struct Foo {
>   #[reflect(ignore)]
>   data: DynamicStruct
> }
> ```

This is already an issue with `FromReflect` (hence why we need to opt-out of it). I think the fix would be something like #9046.

MrGVSV commented:
Rebased and added a new commit (0ada41186f2fac6b7773948d98056c39a82b2e85) based on a [Discord suggestion](https://discord.com/channels/691052431525675048/1002362493634629796/1210316411801706596) from @soqb that re-allows dynamic types in active fields.

Before this commit, this PR would have made the following impossible:

```rust
#[derive(Reflect)]
#[reflect(from_reflect = false)]
struct Foo {
  data: DynamicStruct
}
```

And would have required users to `#reflect(ignore)]` the dynamic field (see https://github.com/bevyengine/bevy/pull/5781#issuecomment-1781964531).

However, this new commit adds a hidden wrapper trait around `GetTypeRegistration`. This allows us to avoid implementing `GetTypeRegistration` for dynamic types while still enabling all type dependency registrations to go through a single function.

james7132 commented on crates/bevy_reflect/src/lib.rs:
```diff
@@ -534,9 +537,48 @@ pub use erased_serde;
 
 extern crate alloc;
 
+/// Exorts used by the reflection macros.
```
```suggestion
/// Exports used by the reflection macros.
```

james7132 reviewed (approved):
Having seen so many PRs come in asking to register one type or another, it's pretty clear that we cannot afford to not include some form of recursive type registration of some kind.

I really am not happy with the manual implementation on RenderAssetUsages, but the rest looks straightforward to me. LGTM.

MrGVSV commented:
> Having seen so many PRs come in asking to register one type or another, it's pretty clear that we cannot afford to not include some form of recursive type registration of some kind.

Agreed!

> I really am not happy with the manual implementation on RenderAssetUsages, but the rest looks straightforward to me. LGTM.

Yeah I'm not sure why `RenderAssetUsages` is implemented manually, but my guess is that it could be replaced by a macro (#12025). But I chose not to mess with it in this PR.

alice-i-cecile commented on crates/bevy_reflect/src/type_registry.rs:
```diff
@@ -100,24 +107,54 @@ impl TypeRegistry {
+    /// To forcibly register the type, overwriting any existing registration, use the
+    /// [`force_add_registration`](Self::force_add_registration) method instead.
+    ///
+    /// Returns `true` if the registration was successfully added,
```
This feels like a weird convention: why not return an `Option<&TypeRegistration>`, with a reference to the previous type registration if it exists?

soqb replied:
agreed.

alice-i-cecile reviewed (approved):
I think this is the right choice: registering types is far too error prone and tedious currently.  I'm generally happy with the code quality.

soqb commented on crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:
```diff
@@ -470,7 +470,12 @@ impl<'a> ReflectMeta<'a> {
+            self,
+            where_clause_options,
+            None,
+            None::<std::iter::Empty<&Type>>,
```
nit: i prefer the following:
```suggestion
            Option::<std::iter::Empty<&Type>>::None,
```

soqb commented on crates/bevy_reflect/src/lib.rs:
```diff
@@ -1903,7 +2063,7 @@ bevy_reflect::tests::Test {
         #[derive(Reflect)]
         #[reflect(where T: Default)]
-        #[reflect(where U: std::ops::Add<T>)]
+        #[reflect(where U: std::ops::Add < T >)]
```
```suggestion
        #[reflect(where U: std::ops::Add<T>)]
```

soqb commented on crates/bevy_reflect/src/type_registry.rs:
````diff
@@ -100,24 +107,54 @@ impl TypeRegistry {
+    /// Additionally, this will add the reflect [data](TypeData) as specified in the [`Reflect`] derive:
     /// ```ignore (Neither bevy_ecs nor serde "derive" are available.)
-    /// #[derive(Component, serde::Serialize, serde::Deserialize, Reflect)]
+    /// #[derive(Reflect)]
````
is this just for the simplicity of the example? if so, i'm on board.

MrGVSV replied:
No haha it wasn't. I probably left it out during a rebase, but yeah I don't think it takes away from what the example is trying to say.

soqb commented on crates/bevy_reflect/src/type_registry.rs:
```diff
@@ -100,24 +107,54 @@ impl TypeRegistry {
+    ///
+    /// To avoid overwriting existing registrations, it's recommended to use the
+    /// [`register`](Self::register) or [`add_registration`](Self::add_registration) methods instead.
+    pub fn force_add_registration(&mut self, registration: TypeRegistration) {
```
nit: why not just `overwrite_registration`?

soqb commented on crates/bevy_reflect_compile_fail_tests/tests/reflect_derive/generics.fail.stderr:
```diff
@@ -1,16 +1,34 @@
+    = help: the following other types implement trait `Reflect`:
+              bool
+              char
+              isize
+              i8
+              i16
+              i32
+              i64
+              i128
+            and $N others
+note: required by a bound in `bevy_reflect::GetField::get_field`
+   --> /home/runner/work/bevy/bevy/crates/bevy_reflect/src/struct_trait.rs:242:21
+    |
+242 |     fn get_field<T: Reflect>(&self, name: &str) -> Option<&T>;
+    |                     ^^^^^^^ required by this bound in `GetField::get_field`
+
+error[E0277]: the trait bound `NoReflect: GetTypeRegistration` is not satisfied
+  --> tests/reflect_derive/generics.fail.rs:14:36
    |
-14 |     let mut foo: Box<dyn Reflect> = Box::new(Foo::<NoReflect> { a: NoReflect(42.0) });
-   |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Reflect` is not implemented for `NoReflect`
+14 |     let mut foo: Box<dyn Struct> = Box::new(Foo::<NoReflect> { a: NoReflect(42.0) });
+   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `GetTypeRegistration` is not implemented for `NoReflect`
```
(non-blocking) annoying that we're lowering the utility of our errors here. i think i ran into this with `TypePath` but i can't remember how i solved it.

MrGVSV replied:
Yeah I wish we had a way of controlling what errors read so that they could be more useful. Not just for bevy_reflect but lots of other parts of Bevy (*looks at `IntoSystemConfig`*).

soqb reviewed (approved):
LGTM! just a few nits.

MrGVSV commented on crates/bevy_reflect/src/type_registry.rs:
```diff
@@ -100,24 +107,57 @@ impl TypeRegistry {
+        &mut self,
+        registration: TypeRegistration,
+    ) -> Option<&TypeRegistration> {
+        if self.contains(registration.type_id()) {
+            self.get(registration.type_id())
+        } else {
+            self.overwrite_registration(registration);
+            None
         }
```
@alice-i-cecile @soqb For some reason I was having a lot of issues getting the borrow checker to be happy with me haha. I don't love that it's adding a second lookup, but it's the best I could come up with right now. Let me know if there's a smarter way fo doing this lol

james7132 commented on crates/bevy_reflect/src/type_registry.rs:
```diff
@@ -100,24 +107,57 @@ impl TypeRegistry {
+        &mut self,
+        registration: TypeRegistration,
+    ) -> Option<&TypeRegistration> {
+        if self.contains(registration.type_id()) {
+            self.get(registration.type_id())
+        } else {
+            self.overwrite_registration(registration);
```
```suggestion
        match self.registrations.entry(registration.type_id()) {
            Entry::Occupied(entry) => {
                 let (_, old) = entry.replace_entry(registration);
                 Some(old)
            }
            Entry::Vacant(entry) => {
                 entry.insert(registration);
                 None
            }
        }
```
@MrGVSV this should work (sans formatting and typos).

MrGVSV replied:
Hm, I think the problem with this is that we don't want to replace the old value. This function shouldn't overwrite.

james7132 replied:
If that's the case, just flip the branches:

```suggestion
        match self.registrations.entry(registration.type_id()) {
            Entry::Occupied(entry) => None,
            Entry::Vacant(entry) => Some(&*entry.insert(registration),
```

MrGVSV replied:
I think this is still flipped. When the entry is `Vacant`, we want to insert the registration and return `None`. When it's `Occupied` we don't want to insert the registration, and instead return `Some(existing)`.

james7132 replied:
This should work then.

```suggestion
        match self.registrations.entry(registration.type_id()) {
            Entry::Occupied(entry) => Some(entry.get()),
            Entry::Vacant(entry) => {
                 entry.insert(registration);
                 None
            },
```

MrGVSV replied:
I’m not at my computer right now but I think when I tried that it gave me an error about returning a reference to a local variable (`entry`). Is it giving it for you as well?

james7132 replied:
Oh I didn't realize that a `get` takes a reference, you may need to use `into_mut` and then downgrade it to a read-only reference.

MrGVSV replied:
Okay so this works, but I forgot that it actually needs to call `self.overwrite_registration` to compute everything and not just `entry.insert`, which means we run into borrow issues again 😵‍💫

james7132 replied:
Ah nevermind then. This can be handled in a separate PR.

cart replied:
Given that my only major concern with this approach has been startup performance for a method that will get called hundreds-to-thousands of times, I think it is worth frontloading this refactor. I've queued up some changes.

SkiFire13 commented on examples/reflection/generic_reflection.rs:
```diff
@@ -12,8 +12,10 @@ fn main() {
         .run();
 }
 
+/// The `#[derive(Reflect)]` macro will automatically add any required bounds to `T`,
+/// such as `Reflect` and `GetTypeRegistration`.
```
Is there a way to opt out of this behaviour (e.g. if you have a `PhantomData<T>`)? If so, it may be nice to also mention that here.

soqb replied:
this is done by default.

MrGVSV replied:
Actually, this example is a bit misleading. The reflection bounds are only added to active fields now. Generic type parameters only get the `TypePath + Send + Sync + 'static` bounds I believe.

So the actual way to opt-out would be to `#[reflect(ignore)]` the field or add `#[reflect(no_field_bounds)]` (probably in combination with `#[reflect(where)]`). 

But I think that's outside the scope of this PR and should probably be tackled on its own since it might require renaming/changing up the example itself.

SkiFire13 replied:
I see, yeah it does get more complex than what should be in this example. Still, the wording of the docs make it seems like there's something not intuitive going on.

I wouldn't consider this a blocker for the PR, though it would be nice if this can be improved now or in the future.

SkiFire13 commented on crates/bevy_reflect/src/type_registry.rs:
```diff
@@ -100,24 +107,86 @@ impl TypeRegistry {
-    pub fn add_registration(&mut self, registration: TypeRegistration) {
-        if self.registrations.contains_key(&registration.type_id()) {
-            return;
+    /// Attempts to register the type described by `registration`.
+    ///
+    /// If the registration for the type already exists, it will not be registered again.
+    /// This also means that its type dependencies will not be registered.
+    ///
+    /// Recursive type dependencies will not be registered more than once.
+    ///
+    /// To forcibly register the type, overwriting any existing registration, use the
+    /// [`overwrite_registration`](Self::overwrite_registration) method instead.
```
The bits about type dependencies are inaccurate. `add_registration` (and `overwrite_registration`) never register type dependencies.

MrGVSV replied:
Oops, that's what I get for writing docs too fast. Good catch!

SkiFire13 commented on crates/bevy_reflect/src/type_registry.rs:
```diff
@@ -100,24 +107,86 @@ impl TypeRegistry {
 
+    /// Registers the type described by `registration`.
+    ///
+    /// If the registration for the type already exists, it will be overwritten.
```
Like for `add_registration` this should mention it never registering type dependencies.

cart reviewed (commented):
My only real concern around this has been startup performance, as we're adding  N TypeId hashes to every app startup, where N is the number of fields across all registered types (which could be hundreds to thousands of hashes). Likely not a huge deal, but it would be good see if it shows up measurably when booting up a default Bevy app.

Given that this is my only concern, I've decided to frontload optimizing the `register` method (instead of doing it in another PR as the current comments decided).

cart reviewed (approved):
I just pushed my register() optimization changes. Please review them before merging.

MrGVSV commented on crates/bevy_reflect/src/type_registry.rs:
```diff
@@ -100,39 +107,114 @@ impl TypeRegistry {
         T: GetTypeRegistration,
     {
-        self.add_registration(T::get_type_registration());
+        if self.add_registration(T::get_type_registration()) {
+            T::register_type_dependencies(self);
+        }
```
@cart Would this possibly incur more `TypeId` hashes than the [previous version](https://github.com/bevyengine/bevy/blob/799a23a93a000b4fb764f0984cff3f75b2544b3d/crates/bevy_reflect/src/type_registry.rs#L151-L156)?

This will call `T::get_type_registration()` every time. And most `GetTypeRegistration` impls will have at least 1 `TypeId` hash via `TypeRegistration::insert` (`ReflectFromPtr`) but more often at least 2 (`ReflectFromReflect`). 

That's the minimum, but most types are registering other type data, such as `ReflectComponent`, `ReflectDefault`, `ReflectSerialize`, etc.

So each call to `register` now will typically require a few extra hashes rather than just two.

cart replied:
Ew yeah wasn't thinking about the cost of `get_type_registration()` at all.

I think this would work:

1. Rework add_registration to accept TypeId and a `fn() -> TypeRegistration`. Call the `fn` only when actually needed internally
2. Call TypeId::of::<T>() directly in `register()`

SkiFire13 replied:
I agree, eagerly computing `T::get_type_registration` is likely not good for performance. Maybe it would be better to add a variant of `add_registration` that takes a closure and only calls it if the registration is not already present?

MrGVSV replied:
I'll try it out. I think it would make sense to make this a private function so we don't allow `add_registration` to be abused (i.e. registering types that are not allowed to be registered).

MrGVSV replied:
Okay I [added `register_internal`](https://github.com/bevyengine/bevy/pull/5781/commits/e9faa7961ddfa4b207534b6e54159a0bbc6befa1) to do just that!

alice-i-cecile commented:
@MrGVSV say the word and I'll merge this in :)

MrGVSV commented:
> @MrGVSV say the word and I'll merge this in :) @SkiFire13 or @soqb your review on the last batch of changes would be nice too.

I mean I’m happy with it haha. We could maybe schedule it to be merged as part of the merge train tomorrow in case anyone has any last minute concerns?

alice-i-cecile commented:
Great: removing the controversial label for now so I don't miss it 😅

james7132 commented:
One final thing that came to mind was maybe to annotate `register_type_dependencies` as `#[inline(never)]` to avoid excessive code generation at compile time.

MrGVSV commented:
> One final thing that came to mind was maybe to annotate `register_type_dependencies` as `#[inline(never)]` to avoid excessive code generation at compile time.

Annotate just the default trait method or add it to the generated `GetTypeRegistration` impls?

james7132 commented:
> @james7132 Annotate just the default trait method or add it to the generated `GetTypeRegistration` impls?

The empty default is fine, but the ones that have an actual body need to be annotated to avoid inlining the recursion.

SkiFire13 commented:
LGTM, although the `#[inline(never)]` feels like premature optimizations. Is it certain that LLVM will inline that method normally (and that this will actually be bad, since it seems it's only used in one function)?
